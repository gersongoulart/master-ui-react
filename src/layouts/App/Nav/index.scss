@import '~variables';

#app[data-viewport="mini"] #container,
#app[data-viewport="tiny"] #container {
    font-size: 100%;
}

#container {
    &.collapsed {
        font-size: 100%;
    }
}

//colors
$color_celeste_approx: transparent;
$black: #000;
$color_pumice_approx: transparent;
$color_mountain_mist_approx:transparent;
$color_storm_dust_approx: transparent;

//@extend-elements
//original selectors
//.mainmenu, .submenu

.main {
    >.nav {
        ul {
            list-style: none;
            padding: 0;
            margin: 0;
        }

        .menu {
            >div {
                .parentMenu {
                    .submenu {
                        overflow: hidden;
                        max-height: 0;
                        transition: all 0.5s ease-out;
                        margin-right: -1em;

                        a {
                            background-color: $color_mountain_mist_approx;
                            padding-left: 0;

                            label {
                                &::before {
                                    padding-left: 10px;
                                    content: '';
                                }
                            }

                            &:last-child {
                                padding-bottom: 0;
                            }

                            &:hover {
                                background-color: $color_storm_dust_approx;
                            }
                        }
                    }

                    &:hover,
                    &.active {
                        .submenu {
                            margin-top: 10px;
                            display: block;
                            max-height: 200px;
                        }
                    }
                }
            }
        }
    }
}
