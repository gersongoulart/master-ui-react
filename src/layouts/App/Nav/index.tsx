/* https://medialoot.com/blog/how-to-create-a-responsive-navigation-menu-using-only-css/ */
/* https://codepen.io/wanni/pen/zsDJb */
import * as React from 'react';
import { Link } from 'react-router';
import { Menu, Icon } from 'semantic-ui-react';
import IRoute from 'core/interfaces/IRoute';
import './index.scss';

interface Props {
    routes: Array<IRoute>;
    position: string;
    activeRoute: string;
}

const Nav: React.StatelessComponent<Props> = (props: Props): JSX.Element => {
    const { routes, position = 'vertical', activeRoute } = props;

    const toggleSubmenu = ({target}) => {
        const isParent = target.classList.contains('parentLevel');
        const el = isParent ? target : target.closest('.parentLevel');
        
        if (!el) {
            return;
        }

        // if (isParent) {
        //     el.classList.toggle('expanded');
        // } 
        // else {
        //     if (!el.classList.contains('expanded')) {
        //         el.classList.add('expanded');
        //     }
        // }
    };   
    
    const getMenuItem = ({ id, href, label, icon, children = [], isChild = false }) => {
        const hasChildren = children.length > 0;
        const Tag: any = Link; 
        const props = Object.assign({}, !hasChildren ? { to: href } : undefined);
        
        return (<li key={id} id={`mnu-${id}`} onClick={hasChildren || isChild ? toggleSubmenu : null} className={`${href === activeRoute ? 'active' : ''} ${hasChildren ? 'parentLevel' : ''}`.trim()}>
            <Tag {...props}>
                { icon && !isChild && <Icon name={ icon } /> }
                { label }
                { hasChildren && <span className="expander"><Icon className="chevron right"></Icon></span>}
            </Tag>
            { hasChildren && <ul className="submenu">{<li className="placeholder"></li>}{children.map((item) => getMenuItem(Object.assign({}, item, {isChild: true})))}</ul> }
        </li>);
    };
    
    const nav = routes.map(getMenuItem);
  
    return <ul className={`menu ${position}`}>{nav}</ul>;
};

export default Nav;
