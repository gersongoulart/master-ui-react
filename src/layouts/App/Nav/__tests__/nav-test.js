/* eslint-env node, jest, mocha, jsx */
import * as React from 'react';
import * as Renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Nav from '../index';
import * as menu from '../../../menu.json';

const defaultProps = { 
    routes: menu['items'],
    activeRoute: '',
    position: 'vertical',
};

function setup (props, test) {
    return () => test(
        Renderer.create(
            <Nav {...props} />
        )
    );
}

describe('LeftPanel Component Rendering', () => {
	it('Should exist', () => {
		expect(Nav).not.toBeNull();
	});

	it('Should render something if it receives required props', setup(
		defaultProps,
		(component) => expect(component.toJSON()).toMatchSnapshot()
	));	
});

describe('LeftPanel behaviour', () => {
    it('should pass a selected value to the onClick handler', setup(
        defaultProps,
		() => {
            const wrapper = shallow(<Nav {...defaultProps} />);
            expect(wrapper).toMatchSnapshot();
        }
	));    
})