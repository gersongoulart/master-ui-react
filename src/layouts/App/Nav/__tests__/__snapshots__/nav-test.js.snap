// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftPanel Component Rendering Should render something if it receives required props 1`] = `
<ul
  className="menu vertical"
>
  <li
    className=""
    id="mnu-dashboard"
    onClick={null}
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <i
        aria-hidden="true"
        className="dashboard icon"
      />
      Dashboard
    </a>
  </li>
  <li
    className="active parentLevel"
    id="mnu-geography"
    onClick={[Function]}
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <i
        aria-hidden="true"
        className="map icon"
      />
      Locations
      <i
        aria-hidden="true"
        className="icon chevron down expander"
      />
    </a>
    <ul
      className="submenu"
    >
      <li
        className="placeholder"
      />
      <li
        className=""
        id="mnu-countries"
        onClick={[Function]}
      >
        <a
          onClick={[Function]}
          style={Object {}}
        >
          Countries
        </a>
      </li>
      <li
        className=""
        id="mnu-city"
        onClick={[Function]}
      >
        <a
          onClick={[Function]}
          style={Object {}}
        >
          Provinces
        </a>
      </li>
    </ul>
  </li>
  <li
    className=""
    id="mnu-notebook"
    onClick={null}
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <i
        aria-hidden="true"
        className="book icon"
      />
      Notebook
    </a>
  </li>
  <li
    className=""
    id="mnu-reports"
    onClick={null}
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <i
        aria-hidden="true"
        className="calculator icon"
      />
      Reports
    </a>
  </li>
</ul>
`;

exports[`LeftPanel behaviour should pass a selected value to the onClick handler 1`] = `
<ul
  className="menu vertical"
>
  <li
    className=""
    id="mnu-dashboard"
    onClick={null}
  >
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/"
    >
      <Icon
        as="i"
        name="dashboard"
      />
      Dashboard
    </Link>
  </li>
  <li
    className="active parentLevel"
    id="mnu-geography"
    onClick={[Function]}
  >
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
    >
      <Icon
        as="i"
        name="map"
      />
      Locations
      <Icon
        as="i"
        className="chevron down expander"
      />
    </Link>
    <ul
      className="submenu"
    >
      <li
        className="placeholder"
      />
      <li
        className=""
        id="mnu-countries"
        onClick={[Function]}
      >
        <Link
          onlyActiveOnIndex={false}
          style={Object {}}
          to="/countries"
        >
          Countries
        </Link>
      </li>
      <li
        className=""
        id="mnu-city"
        onClick={[Function]}
      >
        <Link
          onlyActiveOnIndex={false}
          style={Object {}}
          to="/provinces"
        >
          Provinces
        </Link>
      </li>
    </ul>
  </li>
  <li
    className=""
    id="mnu-notebook"
    onClick={null}
  >
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/notebook"
    >
      <Icon
        as="i"
        name="book"
      />
      Notebook
    </Link>
  </li>
  <li
    className=""
    id="mnu-reports"
    onClick={null}
  >
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/reports"
    >
      <Icon
        as="i"
        name="calculator"
      />
      Reports
    </Link>
  </li>
</ul>
`;
